### PR TITLE
Add Kubernetes Cluster table. closes #44

### LINKED ISSUE #44: Add Kubernetes table
**Describe the new table**

Need to query the Kubernetes table to query if the instances are enabled with RBAC.

**References**

Section 8

We need to achieve some of the steps as per CIS recommendation
8.5 Enable role-based access control (RBAC) within Azure Kubernetes Services (Automated), steps 2 & 3

### PR DESCRIPTION
**Pull request by sayan133**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
No env file present for the current environment:  staging 
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT undefined

SETUP: tests/azure_kubernetes_cluster []

PRETEST: tests/azure_kubernetes_cluster

TEST: tests/azure_kubernetes_cluster
Running terraform
azurerm_resource_group.named_test_resource: Creating...
azurerm_resource_group.named_test_resource: Creation complete after 4s [id=/subscriptions/cd4401c4-3cc8-4565-a594-839c1e345f1e/resourceGroups/turbottest2458]
azurerm_kubernetes_cluster.named_test_resource: Creating...
azurerm_kubernetes_cluster.named_test_resource: Still creating... [10s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [20s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [30s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [40s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [50s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [1m0s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [1m10s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [1m20s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [1m30s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [1m40s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [1m50s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [2m0s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [2m10s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [2m20s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [2m30s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [2m40s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [2m50s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [3m0s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [3m10s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [3m20s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [3m30s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [3m40s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [3m50s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [4m0s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [4m10s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [4m20s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Still creating... [4m30s elapsed]
azurerm_kubernetes_cluster.named_test_resource: Creation complete after 4m36s [id=/subscriptions/cd4401c4-3cc8-4565-a594-839c1e345f1e/resourcegroups/turbottest2458/providers/Microsoft.ContainerService/managedClusters/turbottest2458]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 2 added, 0 changed, 0 destroyed.

Outputs:

cluster_fqdn = "turbottest2458-5fc44aa7.hcp.westus.azmk8s.io"
location = "westus"
node_resource_group = "MC_turbottest2458_turbottest2458_westus"
resource_aka = "azure:///subscriptions/cd4401c4-3cc8-4565-a594-839c1e345f1e/resourcegroups/turbottest2458/providers/Microsoft.ContainerService/managedClusters/turbottest2458"
resource_aka_lower = "azure:///subscriptions/cd4401c4-3cc8-4565-a594-839c1e345f1e/resourcegroups/turbottest2458/providers/microsoft.containerservice/managedclusters/turbottest2458"
resource_id = "/subscriptions/cd4401c4-3cc8-4565-a594-839c1e345f1e/resourcegroups/turbottest2458/providers/Microsoft.ContainerService/managedClusters/turbottest2458"
resource_name = "turbottest2458"
subscription_id = "cd4401c4-3cc8-4565-a594-839c1e345f1e"

Running SQL query: test-get-query.sql
[
  {
    "addon_profiles": {
      "aciConnectorLinux": {
        "enabled": false
      },
      "azurepolicy": {
        "enabled": false
      },
      "httpApplicationRouting": {
        "enabled": false
      },
      "ingressApplicationGateway": {
        "enabled": false
      },
      "kubeDashboard": {
        "enabled": false
      },
      "omsagent": {
        "enabled": false
      }
    },
    "akas": [
      "azure:///subscriptions/cd4401c4-3cc8-4565-a594-839c1e345f1e/resourcegroups/turbottest2458/providers/Microsoft.ContainerService/managedClusters/turbottest2458",
      "azure:///subscriptions/cd4401c4-3cc8-4565-a594-839c1e345f1e/resourcegroups/turbottest2458/providers/microsoft.containerservice/managedclusters/turbottest2458"
    ],
    "api_server_access_profile": {
      "enablePrivateCluster": false
    },
    "dns_prefix": "turbottest2458",
    "enable_rbac": false,
    "fqdn": "turbottest2458-5fc44aa7.hcp.westus.azmk8s.io",
    "id": "/subscriptions/cd4401c4-3cc8-4565-a594-839c1e345f1e/resourcegroups/turbottest2458/providers/Microsoft.ContainerService/managedClusters/turbottest2458",
    "identity": {
      "type": "SystemAssigned"
    },
    "location": "westus",
    "max_agent_pools": 100,
    "name": "turbottest2458",
    "node_resource_group": "MC_turbottest2458_turbottest2458_westus",
    "power_state": {
      "code": "Running"
    },
    "provisioning_state": "Succeeded",
    "region": "westus",
    "resource_group": "turbottest2458",
    "sku": {
      "name": "Basic",
      "tier": "Free"
    },
    "subscription_id": "cd4401c4-3cc8-4565-a594-839c1e345f1e",
    "tags": {
      "foo": "bar"
    },
    "title": "turbottest2458",
    "type": "Microsoft.ContainerService/ManagedClusters"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "id": "/subscriptions/cd4401c4-3cc8-4565-a594-839c1e345f1e/resourcegroups/turbottest2458/providers/Microsoft.ContainerService/managedClusters/turbottest2458",
    "name": "turbottest2458",
    "resource_group": "turbottest2458",
    "type": "Microsoft.ContainerService/ManagedClusters"
  }
]
✔ PASSED

Running SQL query: test-not-found-query.sql
null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "akas": [
      "azure:///subscriptions/cd4401c4-3cc8-4565-a594-839c1e345f1e/resourcegroups/turbottest2458/providers/Microsoft.ContainerService/managedClusters/turbottest2458",
      "azure:///subscriptions/cd4401c4-3cc8-4565-a594-839c1e345f1e/resourcegroups/turbottest2458/providers/microsoft.containerservice/managedclusters/turbottest2458"
    ],
    "name": "turbottest2458",
    "tags": {
      "foo": "bar"
    },
    "title": "turbottest2458"
  }
]
✔ PASSED

POSTTEST: tests/azure_kubernetes_cluster

TEARDOWN: tests/azure_kubernetes_cluster

SUMMARY:

1/1 passed.

```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
### Basic Info

```sql
select
  name,
  id,
  location,
  type,
  sku
from
  azure_kubernetes_cluster;
```
```
+----------------+--------------------------------------------------------------------------------------------------------------------------------------------------+----------+--------------------------------------------+------------------------+
| name           | id                                                                                                                                               | location | type                                       | sku                    |
+----------------+--------------------------------------------------------------------------------------------------------------------------------------------------+----------+--------------------------------------------+------------------------+
| test-steampipe | /subscriptions/cd4401c4-3cc8-4565-a594-839c1e345f1e/resourcegroups/turbot_rg/providers/Microsoft.ContainerService/managedClusters/test-steampipe | westus   | Microsoft.ContainerService/ManagedClusters | {Name:Basic Tier:Free} |
+----------------+--------------------------------------------------------------------------------------------------------------------------------------------------+----------+--------------------------------------------+------------------------+
```

### List Kubernetes Clusters for a specific location

```sql
select
  name,
  id,
  location,
  type,
  identity,
  sku
from
  azure_kubernetes_cluster
where
  location = 'westus';
```

```
+----------------+--------------------------------------------------------------------------------------------------------------------------------------------------+----------+--------------------------------------------+---------------------------+------------------------+
| name           | id                                                                                                                                               | location | type                                       | identity                  | sku                    |
+----------------+--------------------------------------------------------------------------------------------------------------------------------------------------+----------+--------------------------------------------+---------------------------+------------------------+
| test-steampipe | /subscriptions/cd4401c4-3cc8-4565-a594-839c1e345f1e/resourcegroups/turbot_rg/providers/Microsoft.ContainerService/managedClusters/test-steampipe | westus   | Microsoft.ContainerService/ManagedClusters | {"type":"SystemAssigned"} | {Name:Basic Tier:Free} |
+----------------+--------------------------------------------------------------------------------------------------------------------------------------------------+----------+--------------------------------------------+---------------------------+------------------------+
```

### List Kubernetes Clusters having System Assigned identity

```sql
select
  name,
  id,
  location,
  type,
  identity ->> 'type' as identity_type,
  sku
from
  azure_kubernetes_cluster
where
  identity ->> 'type' = 'SystemAssigned';
```
```
+----------------+--------------------------------------------------------------------------------------------------------------------------------------------------+----------+--------------------------------------------+----------------+------------------------+
| name           | id                                                                                                                                               | location | type                                       | identity_type  | sku                    |
+----------------+--------------------------------------------------------------------------------------------------------------------------------------------------+----------+--------------------------------------------+----------------+------------------------+
| test-steampipe | /subscriptions/cd4401c4-3cc8-4565-a594-839c1e345f1e/resourcegroups/turbot_rg/providers/Microsoft.ContainerService/managedClusters/test-steampipe | westus   | Microsoft.ContainerService/ManagedClusters | SystemAssigned | {Name:Basic Tier:Free} |
+----------------+--------------------------------------------------------------------------------------------------------------------------------------------------+----------+--------------------------------------------+----------------+------------------------+
```

### List Clusters having role-based access control (RBAC) disabled

```sql
select
  name,
  id,
  location,
  type,
  identity,
  enable_rbac,
  sku
from
  azure_kubernetes_cluster
where
  not enable_rbac;
```
```
+------+----+----------+------+----------+-------------+-----+
| name | id | location | type | identity | enable_rbac | sku |
+------+----+----------+------+----------+-------------+-----+
+------+----+----------+------+----------+-------------+-----+
```

### List Clusters having undesired/older kubernetes version

```sql
select
  name,
  id,
  location,
  type,
  kubernetes_version
from
  azure_kubernetes_cluster
where
  kubernetes_version < '1.20.5';
```
```
+--------+------------------------------------------------------------------------------------------------------------------------------------------+----------+--------------------------------------------+--------------------+
| name   | id                                                                                                                                       | location | type                                       | kubernetes_version |
+--------+------------------------------------------------------------------------------------------------------------------------------------------+----------+--------------------------------------------+--------------------+
| del-me | /subscriptions/cd4401c4-3cc8-4565-a594-839c1e345f1e/resourcegroups/turbot_rg/providers/Microsoft.ContainerService/managedClusters/del-me | westus   | Microsoft.ContainerService/ManagedClusters | 1.19.9             |
+--------+------------------------------------------------------------------------------------------------------------------------------------------+----------+--------------------------------------------+--------------------+

```
</details>

### CODE REVIEW

File: azure/table_azure_kubernetes_cluster.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/containerservice/mgmt/2021-02-01/containerservice"
+	"github.com/turbot/steampipe-plugin-sdk/grpc/proto"
+	"github.com/turbot/steampipe-plugin-sdk/plugin/transform"
+
+	"github.com/turbot/steampipe-plugin-sdk/plugin"
+)
+
+//// TABLE DEFINITION ////
```
Comment (Paulami30) [inline azure/table_azure_kubernetes_cluster.go:14]:
```suggestion
//// TABLE DEFINITION
```

File: docs/tables/azure_kubernetes_cluster.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  id,
+  location,
+  type,
+  sku
+from
+  azure_kubernetes_cluster;
+```
+
+
+### List Kubernetes Clusters for a specific location
```
Comment (Paulami30) [inline docs/tables/azure_kubernetes_cluster.md:21]:
```suggestion
### List clusters for a specific location
```

Comment (Paulami30) [reply docs/tables/azure_kubernetes_cluster.md:21]: Remove service name from all the headings

File: azure/table_azure_kubernetes_cluster.go [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+			Hydrate:           getKubernetesCluster,
+		},
+		List: &plugin.ListConfig{
+			Hydrate: listKubernetesClusters,
+		},
+		Columns: []*plugin.Column{
+			{
+				Name:        "name",
+				Type:        proto.ColumnType_STRING,
+				Description: "The friendly name that identifies the vault",
```
Comment (Paulami30) [inline azure/table_azure_kubernetes_cluster.go:31]:
```suggestion
				Description: "The friendly name that identifies the vault.",
```

File: azure/table_azure_kubernetes_cluster.go [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+				Description: "The type of the cluster.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "location",
+				Description: "The location where the cluster is created.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "sku",
```
Comment (khushboo9024) [inline azure/table_azure_kubernetes_cluster.go:50]: `sku` column should be of type json , please check once.

File: azure/table_azure_kubernetes_cluster.go [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+
+func getKubernetesCluster(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("getKubernetesCluster")
+
+	resourceName := ""
+	resourceGroupName := ""
+	if h.Item != nil {
+		managedCluster := h.Item.(containerservice.ManagedCluster)
+		resourceName = *managedCluster.Name
+		resourceGroupName = strings.Split(string(*managedCluster.ID), "/")[4]
```
Comment (khushboo9024) [inline azure/table_azure_kubernetes_cluster.go:290]: do we have the different response in list and get call ?

File: azure/table_azure_kubernetes_cluster.go [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+			{
+				Name:        "sku",
+				Description: "The managed cluster SKU.",
+				Type:        proto.ColumnType_STRING,
+				Hydrate:     getKubernetesCluster,
+			},
+			{
+				Name:        "identity",
+				Description: "The identity of the managed cluster, if configured.",
+				Type:        proto.ColumnType_JSON,
```
Comment (khushboo9024) [inline azure/table_azure_kubernetes_cluster.go:58]: Json column should be at last.

**Review by khushboo9024 — CHANGES_REQUESTED** [review]
Please make suggested changes.

File: azure/table_azure_kubernetes_cluster.go [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+	"context"
+
+	"github.com/Azure/azure-sdk-for-go/services/containerservice/mgmt/2021-02-01/containerservice"
+	"github.com/turbot/steampipe-plugin-sdk/grpc/proto"
+	"github.com/turbot/steampipe-plugin-sdk/plugin/transform"
+
+	"github.com/turbot/steampipe-plugin-sdk/plugin"
+)
+
+//// TABLE DEFINITION ////
```
Comment (Subhajit97) [inline azure/table_azure_kubernetes_cluster.go:13]:
```suggestion
//// TABLE DEFINITION
```

File: azure/table_azure_kubernetes_cluster.go [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+				Name:        "service_principal_profile",
+				Description: "Information about a service principal identity for the cluster to use for manipulating Azure APIs.",
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("ManagedClusterProperties.ServicePrincipalProfile"),
+			},
+			{
+				Name:        "sku",
+				Description: "The managed cluster SKU.",
+				Type:        proto.ColumnType_JSON,
+				// Hydrate:     getKubernetesCluster,
```
Comment (Subhajit97) [inline azure/table_azure_kubernetes_cluster.go:201]:
```suggestion
```

File: azure/table_azure_kubernetes_cluster.go [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+			},
+			{
+				Name:        "windows_profile",
+				Description: "Profile for Windows VMs in the container service cluster.",
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("ManagedClusterProperties.WindowsProfile"),
+			},
+
+			// Standard columns
```
Comment (Subhajit97) [inline azure/table_azure_kubernetes_cluster.go:210]:
```suggestion
			// Steampipe standard columns
```

File: azure/table_azure_kubernetes_cluster.go [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+				Description: ColumnDescriptionTags,
+				Type:        proto.ColumnType_JSON,
+			},
+			{
+				Name:        "akas",
+				Description: ColumnDescriptionAkas,
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("ID").Transform(idToAkas),
+			},
+			{
```
Comment (Subhajit97) [inline azure/table_azure_kubernetes_cluster.go:228]:
```suggestion
			// Azure standard columns
			{
```

File: azure/table_azure_kubernetes_cluster.go [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+			d.StreamListItem(ctx, cluster)
+		}
+		result.NextWithContext(context.Background())
+		pagesLeft = result.NotDone()
+	}
+
+	return nil, err
+}
+
+//// HYDRATE FUNCTION
```
Comment (Subhajit97) [inline azure/table_azure_kubernetes_cluster.go:279]:
```suggestion
//// HYDRATE FUNCTIONS
```

File: azure/table_azure_kubernetes_cluster.go [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+
+func getKubernetesCluster(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("getKubernetesCluster")
+
+	resourceName := d.KeyColumnQuals["name"].GetStringValue()
+	resourceGroupName := d.KeyColumnQuals["resource_group"].GetStringValue()
```
Comment (Subhajit97) [inline azure/table_azure_kubernetes_cluster.go:285]: Move this code blocks after creating the session

**Review by Subhajit97 — CHANGES_REQUESTED** [review]
@sayan133 There are some minor changes required. Please see comments, thanks!

File: docs/tables/azure_kubernetes_cluster.md [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+  id,
+  location,
+  type,
+  identity,
+  enable_rbac,
+  sku
+from
+  azure_kubernetes_cluster
+where
+  enable_rbac = 'false';
```
Comment (Subhajit97) [inline docs/tables/azure_kubernetes_cluster.md:69]:
```suggestion
  enable_rbac;
```

File: azure/table_azure_kubernetes_cluster.go [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+		result.NextWithContext(context.Background())
+		pagesLeft = result.NotDone()
+	}
+
+	return nil, err
+}
+
+//// HYDRATE FUNCTIONS
+
+func getKubernetesCluster(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
```
Comment (Subhajit97) [inline azure/table_azure_kubernetes_cluster.go:282]:
```suggestion
func getKubernetesCluster(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
```

**Review by Subhajit97 — APPROVED** [review]
LGTM

**Review by rajmohanty17 — CHANGES_REQUESTED** [review]
Can we add / extend the example to see the version / or check the latest version being used or not ?

**Review by rajmohanty17 — APPROVED** [review]
LGTM

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**